File: admin/google_sheets.py
import os
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
import pickle

# Если вы изменяете эти области, удалите файл token.pickle.
SCOPES = ['https://www.googleapis.com/auth/spreadsheets.readonly']

def get_credentials():
    creds = None
    # Файл token.pickle хранит токены доступа и обновления пользователя
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    
    # Если нет действительных учетных данных, позволяем пользователю войти.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
        # Сохраняем учетные данные для следующего запуска
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)
    
    return creds

def get_sheet_service():
    creds = get_credentials()
    service = build('sheets', 'v4', credentials=creds)
    return service.spreadsheets()

def get_nominations():
    """Получение списка номинаций из таблицы"""
    sheet = get_sheet_service()
    spreadsheet_id = os.environ.get('SPREADSHEET_ID')
    range_name = 'NewRes!A:A'
    
    result = sheet.values().get(
        spreadsheetId=spreadsheet_id,
        range=range_name
    ).execute()
    
    values = result.get('values', [])
    if not values:
        return []
    
    # Получаем уникальные номинации из первого столбца
    nominations = set()
    for row in values[1:]:  # Пропускаем заголовок
        if row and row[0]:
            nominations.add(row[0])
    
    return sorted(list(nominations))

def get_blocks():
    """Получение списка блоков из таблицы"""
    sheet = get_sheet_service()
    spreadsheet_id = os.environ.get('SPREADSHEET_ID')
    range_name = 'NewRes!B:B'
    
    result = sheet.values().get(
        spreadsheetId=spreadsheet_id,
        range=range_name
    ).execute()
    
    values = result.get('values', [])
    if not values:
        return []
    
    # Получаем уникальные блоки из второго столбца
    blocks = set()
    for row in values[1:]:  # Пропускаем заголовок
        if row and len(row) > 1 and row[1]:
            blocks.add(row[1])
    
    return sorted(list(blocks))

def fetch_final_scores(nomination=None, block=None):
    """Получение финальных оценок с фильтрацией по номинации и блоку"""
    sheet = get_sheet_service()
    spreadsheet_id = os.environ.get('SPREADSHEET_ID')
    range_name = 'NewRes!A:Z'
    
    result = sheet.values().get(
        spreadsheetId=spreadsheet_id,
        range=range_name
    ).execute()
    
    values = result.get('values', [])
    if not values:
        return []
    
    headers = values[0]
    scores = []
    
    for row in values[1:]:
        if len(row) < len(headers):
            row.extend([''] * (len(headers) - len(row)))
        
        row_data = dict(zip(headers, row))
        
        # Применяем фильтры
        if nomination and row_data.get('Номинация') != nomination:
            continue
        if block and row_data.get('Блок') != block:
            continue
        
        # Собираем оценки судей
        judge_scores = []
        for header in headers:
            if header.startswith('Судья'):
                score = row_data.get(header)
                if score and score.strip():
                    judge_scores.append({
                        'judge': header,
                        'score': float(score)
                    })
        
        if judge_scores:
            scores.append({
                'participant': row_data.get('Участник', ''),
                'scores': judge_scores
            })
    
    return scores

def get_nomination_leaders(nomination, limit=5):
    """Получение лидеров в номинации"""
    scores = fetch_final_scores(nomination)
    
    # Вычисляем средний балл для каждого участника
    participants = {}
    for score in scores:
        participant = score['participant']
        if participant not in participants:
            participants[participant] = {
                'participant': participant,
                'total_score': 0,
                'count': 0
            }
        
        for judge_score in score['scores']:
            participants[participant]['total_score'] += judge_score['score']
            participants[participant]['count'] += 1
    
    # Вычисляем средний балл и сортируем
    leaders = []
    for participant in participants.values():
        if participant['count'] > 0:
            leaders.append({
                'participant': participant['participant'],
                'average_score': participant['total_score'] / participant['count']
            })
    
    leaders.sort(key=lambda x: x['average_score'], reverse=True)
    return leaders[:limit]

def get_participants(block=None):
    """Получение списка участников с фильтрацией по блоку"""
    sheet = get_sheet_service()
    spreadsheet_id = os.environ.get('SPREADSHEET_ID')
    range_name = 'NewRes!A:C'
    
    result = sheet.values().get(
        spreadsheetId=spreadsheet_id,
        range=range_name
    ).execute()
    
    values = result.get('values', [])
    if not values:
        return []
    
    participants = []
    for row in values[1:]:  # Пропускаем заголовок
        if len(row) >= 3:
            participant_block = row[1]
            if block and participant_block != block:
                continue
                
            participants.append({
                'number': row[0],
                'name': row[2],
                'block': participant_block
            })
    
    return participants 

File: admin/google_sheets.js
const fs = require('fs');
const path = require('path');
const { google } = require('googleapis');

// Если вы изменяете эти области, удалите файл token.json
const SCOPES = ['https://www.googleapis.com/auth/spreadsheets.readonly'];

/**
 * Получение и обновление учетных данных для доступа к API
 */
async function getCredentials() {
  let credentials = null;
  const tokenPath = path.join(__dirname, 'token.json');
  
  // Проверяем, есть ли сохраненный токен
  if (fs.existsSync(tokenPath)) {
    const token = JSON.parse(fs.readFileSync(tokenPath, 'utf8'));
    credentials = google.auth.fromJSON(token);
  }
  
  // Если нет действительных учетных данных, запускаем процесс авторизации
  if (!credentials || !credentials.valid) {
    if (credentials && credentials.refreshToken) {
      await credentials.refreshAccessToken();
    } else {
      const credentialsPath = path.join(__dirname, 'credentials.json');
      const content = fs.readFileSync(credentialsPath, 'utf8');
      const { client_secret, client_id, redirect_uris } = JSON.parse(content).installed;
      
      const oAuth2Client = new google.auth.OAuth2(
        client_id, client_secret, redirect_uris[0]
      );
      
      // Открываем браузер для авторизации
      console.log('Authorize this app by visiting this url:');
      const authUrl = oAuth2Client.generateAuthUrl({
        access_type: 'offline',
        scope: SCOPES,
      });
      console.log(authUrl);
      
      // Здесь следует добавить код для получения кода авторизации
      // В реальном приложении нужно реализовать веб-сервер для обработки callback
      // В этом примере мы ожидаем ввод кода пользователем
      const readline = require('readline');
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
      });
      
      const code = await new Promise((resolve) => {
        rl.question('Enter the code from that page here: ', (code) => {
          rl.close();
          resolve(code);
        });
      });
      
      const { tokens } = await oAuth2Client.getToken(code);
      oAuth2Client.setCredentials(tokens);
      credentials = oAuth2Client;
      
      // Сохраняем токен для следующего запуска
      fs.writeFileSync(tokenPath, JSON.stringify(tokens));
    }
  }
  
  return credentials;
}

/**
 * Получение сервиса для работы с Google Sheets
 */
async function getSheetService() {
  const auth = await getCredentials();
  const sheets = google.sheets({ version: 'v4', auth });
  return sheets.spreadsheets;
}

/**
 * Получение списка номинаций из таблицы
 */
async function getNominations() {
  try {
    const sheets = await getSheetService();
    const spreadsheetId = process.env.SPREADSHEET_ID;
    const range = 'NewRes!A:A';
    
    const response = await sheets.values.get({
      spreadsheetId,
      range,
    });
    
    const rows = response.data.values || [];
    if (rows.length === 0) {
      return [];
    }
    
    // Получаем уникальные номинации из первого столбца
    const nominationsSet = new Set();
    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      if (row && row[0]) {
        nominationsSet.add(row[0]);
      }
    }
    
    return Array.from(nominationsSet).sort();
  } catch (error) {
    console.error('Error fetching nominations:', error);
    throw error;
  }
}

/**
 * Получение списка блоков из таблицы
 */
async function getBlocks() {
  try {
    const sheets = await getSheetService();
    const spreadsheetId = process.env.SPREADSHEET_ID;
    const range = 'NewRes!B:B';
    
    const response = await sheets.values.get({
      spreadsheetId,
      range,
    });
    
    const rows = response.data.values || [];
    if (rows.length === 0) {
      return [];
    }
    
    // Получаем уникальные блоки из второго столбца
    const blocksSet = new Set();
    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      if (row && row.length > 0 && row[0]) {
        blocksSet.add(row[0]);
      }
    }
    
    return Array.from(blocksSet).sort();
  } catch (error) {
    console.error('Error fetching blocks:', error);
    throw error;
  }
}

/**
 * Получение финальных оценок с фильтрацией по номинации и блоку
 */
async function fetchFinalScores(nomination = null, block = null) {
  try {
    const sheets = await getSheetService();
    const spreadsheetId = process.env.SPREADSHEET_ID;
    const range = 'NewRes!A:Z';
    
    const response = await sheets.values.get({
      spreadsheetId,
      range,
    });
    
    const rows = response.data.values || [];
    if (rows.length === 0) {
      return [];
    }
    
    const headers = rows[0];
    const scores = [];
    
    for (let i = 1; i < rows.length; i++) {
      let row = rows[i];
      
      // Дополняем строку пустыми значениями, если не хватает столбцов
      if (row.length < headers.length) {
        row = [...row, ...Array(headers.length - row.length).fill('')];
      }
      
      // Создаем объект с данными строки
      const rowData = {};
      for (let j = 0; j < headers.length; j++) {
        rowData[headers[j]] = row[j] || '';
      }
      
      // Применяем фильтры
      if (nomination && rowData['Номинация'] !== nomination) {
        continue;
      }
      if (block && rowData['Блок'] !== block) {
        continue;
      }
      
      // Собираем оценки судей
      const judgeScores = [];
      for (const header of headers) {
        if (header.startsWith('Судья')) {
          const score = rowData[header];
          if (score && score.trim()) {
            judgeScores.push({
              judge: header,
              score: parseFloat(score)
            });
          }
        }
      }
      
      if (judgeScores.length > 0) {
        scores.push({
          participant: rowData['Участник'] || '',
          scores: judgeScores
        });
      }
    }
    
    return scores;
  } catch (error) {
    console.error('Error fetching scores:', error);
    throw error;
  }
}

/**
 * Получение лидеров в номинации
 */
async function getNominationLeaders(nomination, limit = 5) {
  try {
    const scores = await fetchFinalScores(nomination);
    
    // Вычисляем средний балл для каждого участника
    const participants = {};
    for (const score of scores) {
      const participant = score.participant;
      if (!participants[participant]) {
        participants[participant] = {
          participant,
          totalScore: 0,
          count: 0
        };
      }
      
      for (const judgeScore of score.scores) {
        participants[participant].totalScore += judgeScore.score;
        participants[participant].count++;
      }
    }
    
    // Вычисляем средний балл и сортируем
    const leaders = [];
    for (const participantId in participants) {
      const participant = participants[participantId];
      if (participant.count > 0) {
        leaders.push({
          participant: participant.participant,
          averageScore: participant.totalScore / participant.count
        });
      }
    }
    
    leaders.sort((a, b) => b.averageScore - a.averageScore);
    return leaders.slice(0, limit);
  } catch (error) {
    console.error('Error getting nomination leaders:', error);
    throw error;
  }
}

/**
 * Получение списка участников с фильтрацией по блоку
 */
async function getParticipants(block = null) {
  try {
    const sheets = await getSheetService();
    const spreadsheetId = process.env.SPREADSHEET_ID;
    const range = 'NewRes!A:C';
    
    const response = await sheets.values.get({
      spreadsheetId,
      range,
    });
    
    const rows = response.data.values || [];
    if (rows.length === 0) {
      return [];
    }
    
    const participants = [];
    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      if (row.length >= 3) {
        const participantBlock = row[1];
        if (block && participantBlock !== block) {
          continue;
        }
        
        participants.push({
          number: row[0],
          name: row[2],
          block: participantBlock
        });
      }
    }
    
    return participants;
  } catch (error) {
    console.error('Error fetching participants:', error);
    throw error;
  }
}

module.exports = {
  getNominations,
  getBlocks,
  fetchFinalScores,
  getNominationLeaders,
  getParticipants,
  getSheetService
}; 